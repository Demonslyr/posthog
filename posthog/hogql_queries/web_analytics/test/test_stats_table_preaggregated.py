from unittest.mock import patch

import pytest
from posthog.hogql import ast
from posthog.hogql.printer import print_ast
from posthog.hogql.context import HogQLContext
from posthog.hogql.test.utils import pretty_print_in_tests

from posthog.clickhouse.client.execute import sync_execute
from posthog.hogql_queries.web_analytics.stats_table import WebStatsTableQueryRunner
from posthog.hogql_queries.web_analytics.stats_table_pre_aggregated import StatsTablePreAggregatedQueryBuilder
from posthog.models.utils import uuid7
from posthog.schema import (
    WebStatsTableQuery,
    DateRange,
    WebStatsBreakdown,
    WebAnalyticsOrderByFields,
    WebAnalyticsOrderByDirection,
    CompareFilter,
    EventPropertyFilter,
)
from posthog.test.base import (
    APIBaseTest,
    ClickhouseTestMixin,
)


class TestStatsTablePreAggregated(ClickhouseTestMixin, APIBaseTest):
    """
    Test suite for stats table pre-aggregated functionality that verifies
    state aggregations work correctly with the actual query runners.
    """

    snapshot: any

    def _create_events(self, data: list[dict]) -> None:
        """Helper method to create events in ClickHouse for testing."""
        for event in data:
            sync_execute(
                """
                INSERT INTO events (uuid, team_id, distinct_id, timestamp, event, properties, person_id)
                VALUES (%(uuid)s, %(team_id)s, %(distinct_id)s, %(timestamp)s, %(event)s, %(properties)s, %(person_id)s)
                """,
                {
                    "uuid": event.get("uuid", str(uuid7())),
                    "team_id": self.team.pk,
                    "distinct_id": event["distinct_id"],
                    "timestamp": event["timestamp"],
                    "event": event.get("event", "$pageview"),
                    "properties": event.get("properties", {}),
                    "person_id": event.get("person_id", str(uuid7())),
                },
            )

    @pytest.mark.django_db
    def test_can_use_preaggregated_tables_device_breakdown(self):
        """Test that we can use pre-aggregated tables for device breakdown."""
        query = WebStatsTableQuery(
            dateRange=DateRange(date_from="2023-01-01", date_to="2023-01-02"),
            breakdownBy=WebStatsBreakdown.DEVICE_TYPE,
            properties=[],
            limit=10,
        )

        runner = WebStatsTableQueryRunner(team=self.team, query=query)
        builder = StatsTablePreAggregatedQueryBuilder(runner)

        assert builder.can_use_preaggregated_tables() is True

    @pytest.mark.django_db
    def test_cannot_use_preaggregated_for_unsupported_breakdown(self):
        """Test that we cannot use pre-aggregated tables for unsupported breakdowns."""
        query = WebStatsTableQuery(
            dateRange=DateRange(date_from="2023-01-01", date_to="2023-01-02"),
            breakdownBy=WebStatsBreakdown.FRUSTRATION_METRICS,  # Not supported
            properties=[],
            limit=10,
        )

        runner = WebStatsTableQueryRunner(team=self.team, query=query)
        builder = StatsTablePreAggregatedQueryBuilder(runner)

        assert builder.can_use_preaggregated_tables() is False

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_device_type_breakdown_query_generation(self):
        """Test query generation for device type breakdown with state aggregations."""
        query = WebStatsTableQuery(
            dateRange=DateRange(date_from="2023-01-01", date_to="2023-01-02"),
            breakdownBy=WebStatsBreakdown.DEVICE_TYPE,
            properties=[],
            limit=10,
        )

        runner = WebStatsTableQueryRunner(team=self.team, query=query)
        builder = StatsTablePreAggregatedQueryBuilder(runner)

        query_ast = builder.get_query()
        sql = self._print_select(query_ast)

        assert self.snapshot == sql

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_browser_breakdown_query_generation(self):
        """Test query generation for browser breakdown with state aggregations."""
        query = WebStatsTableQuery(
            dateRange=DateRange(date_from="2023-01-01", date_to="2023-01-02"),
            breakdownBy=WebStatsBreakdown.BROWSER,
            properties=[],
            orderBy=(WebAnalyticsOrderByFields.VISITORS, WebAnalyticsOrderByDirection.DESC),
            limit=5,
        )

        runner = WebStatsTableQueryRunner(team=self.team, query=query)
        builder = StatsTablePreAggregatedQueryBuilder(runner)

        query_ast = builder.get_query()
        sql = self._print_select(query_ast)

        assert self.snapshot == sql

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_page_breakdown_with_bounce_rate_query_generation(self):
        """Test query generation for page breakdown with bounce rate calculations using state aggregations."""
        query = WebStatsTableQuery(
            dateRange=DateRange(date_from="2023-01-01", date_to="2023-01-02"),
            breakdownBy=WebStatsBreakdown.PAGE,
            properties=[],
            includeBounceRate=True,
            limit=10,
        )

        runner = WebStatsTableQueryRunner(team=self.team, query=query)
        builder = StatsTablePreAggregatedQueryBuilder(runner)

        query_ast = builder.get_query()
        sql = self._print_select(query_ast)

        assert self.snapshot == sql

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_initial_page_breakdown_with_bounce_rate_query_generation(self):
        """Test query generation for initial page breakdown with bounce rate using state aggregations."""
        query = WebStatsTableQuery(
            dateRange=DateRange(date_from="2023-01-01", date_to="2023-01-02"),
            breakdownBy=WebStatsBreakdown.INITIAL_PAGE,
            properties=[],
            includeBounceRate=True,
            limit=10,
        )

        runner = WebStatsTableQueryRunner(team=self.team, query=query)
        builder = StatsTablePreAggregatedQueryBuilder(runner)

        query_ast = builder.get_query()
        sql = self._print_select(query_ast)

        assert self.snapshot == sql

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_utm_source_breakdown_with_comparison_period(self):
        """Test query generation for UTM source breakdown with comparison period using state aggregations."""
        query = WebStatsTableQuery(
            dateRange=DateRange(date_from="2023-01-08", date_to="2023-01-15"),
            breakdownBy=WebStatsBreakdown.INITIAL_UTM_SOURCE,
            properties=[],
            compareFilter=CompareFilter(compare=True, compare_to="previous_period"),
            limit=10,
        )

        runner = WebStatsTableQueryRunner(team=self.team, query=query)
        builder = StatsTablePreAggregatedQueryBuilder(runner)

        query_ast = builder.get_query()
        sql = self._print_select(query_ast)

        assert self.snapshot == sql

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_country_breakdown_with_filters(self):
        """Test query generation for country breakdown with property filters using state aggregations."""
        query = WebStatsTableQuery(
            dateRange=DateRange(date_from="2023-01-01", date_to="2023-01-02"),
            breakdownBy=WebStatsBreakdown.COUNTRY,
            properties=[EventPropertyFilter(key="$browser", value="Chrome", operator="exact", type="event")],
            limit=10,
        )

        runner = WebStatsTableQueryRunner(team=self.team, query=query)
        builder = StatsTablePreAggregatedQueryBuilder(runner)

        query_ast = builder.get_query()
        sql = self._print_select(query_ast)

        assert self.snapshot == sql

    @pytest.mark.usefixtures("unittest_snapshot")
    @patch(
        "posthog.hogql_queries.web_analytics.stats_table_pre_aggregated.StatsTablePreAggregatedQueryBuilder.can_use_preaggregated_tables"
    )
    def test_fallback_to_regular_query_when_preaggregated_not_available(self, mock_can_use):
        """Test that the runner falls back to regular queries when pre-aggregated tables can't be used."""
        mock_can_use.return_value = False

        query = WebStatsTableQuery(
            dateRange=DateRange(date_from="2023-01-01", date_to="2023-01-02"),
            breakdownBy=WebStatsBreakdown.DEVICE_TYPE,
            properties=[],
            limit=10,
        )

        # Disable pre-aggregated tables to test fallback
        runner = WebStatsTableQueryRunner(team=self.team, query=query)

        query_ast = runner.to_query()
        sql = self._print_select(query_ast)

        # Should use the regular query path (with uniq, count instead of merge functions)
        assert self.snapshot == sql

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_complex_stats_table_with_state_aggregations(self):
        """Test complex stats table query combining multiple state aggregation functions."""
        query = WebStatsTableQuery(
            dateRange=DateRange(date_from="2023-01-01", date_to="2023-01-07"),
            breakdownBy=WebStatsBreakdown.VIEWPORT,
            properties=[],
            compareFilter=CompareFilter(compare=True, compare_to="previous_period"),
            orderBy=(WebAnalyticsOrderByFields.VIEWS, WebAnalyticsOrderByDirection.DESC),
            limit=5,
        )

        runner = WebStatsTableQueryRunner(team=self.team, query=query)
        builder = StatsTablePreAggregatedQueryBuilder(runner)

        query_ast = builder.get_query()
        sql = self._print_select(query_ast)

        # Verify it uses merge functions for state aggregations
        assert "uniqMergeIf" in sql
        assert "sumMergeIf" in sql
        assert self.snapshot == sql

    def _print_select(self, expr: ast.SelectQuery | ast.SelectSetQuery):
        """Helper method to print AST as formatted SQL."""
        query = print_ast(
            expr,
            HogQLContext(team_id=self.team.pk, enable_select_queries=True),
            "clickhouse",
        )
        return pretty_print_in_tests(query, self.team.pk)
