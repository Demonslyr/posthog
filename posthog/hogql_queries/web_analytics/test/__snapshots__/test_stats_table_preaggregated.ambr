# serializer version: 1
# name: TestStatsTablePreAggregated.test_browser_breakdown_query_generation
  '''
  
  SELECT web_stats_daily.browser AS `context.columns.breakdown_value`, tuple(uniqMergeIf(web_stats_daily.persons_uniq_state, and(greaterOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-02 23:59:59.999999', 6, 'UTC')))), uniqMergeIf(web_stats_daily.persons_uniq_state, and(greaterOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-02 23:59:59.999999', 6, 'UTC'))))) AS `context.columns.visitors`, tuple(sumMergeIf(web_stats_daily.pageviews_count_state, and(greaterOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-02 23:59:59.999999', 6, 'UTC')))), sumMergeIf(web_stats_daily.pageviews_count_state, and(greaterOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-02 23:59:59.999999', 6, 'UTC'))))) AS `context.columns.views` 
  FROM web_stats_daily FINAL 
  WHERE and(equals(web_stats_daily.team_id, 420), greaterOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-02 23:59:59.999999', 6, 'UTC'))) 
  GROUP BY `context.columns.breakdown_value` ORDER BY `context.columns.visitors` DESC 
  LIMIT 50000
  '''
# ---
# name: TestStatsTablePreAggregated.test_complex_stats_table_with_state_aggregations
  '''
  
  SELECT concat(ifNull(toString(web_stats_daily.viewport_width), ''), %(hogql_val_0)s, ifNull(toString(web_stats_daily.viewport_height), '')) AS `context.columns.breakdown_value`, tuple(uniqMergeIf(web_stats_daily.persons_uniq_state, and(greaterOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-07 23:59:59.999999', 6, 'UTC')))), uniqMergeIf(web_stats_daily.persons_uniq_state, and(greaterOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-07 23:59:59.999999', 6, 'UTC'))))) AS `context.columns.visitors`, tuple(sumMergeIf(web_stats_daily.pageviews_count_state, and(greaterOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-07 23:59:59.999999', 6, 'UTC')))), sumMergeIf(web_stats_daily.pageviews_count_state, and(greaterOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-07 23:59:59.999999', 6, 'UTC'))))) AS `context.columns.views` 
  FROM web_stats_daily FINAL 
  WHERE and(equals(web_stats_daily.team_id, 420), greaterOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-07 23:59:59.999999', 6, 'UTC'))) 
  GROUP BY `context.columns.breakdown_value` ORDER BY `context.columns.views` DESC 
  LIMIT 50000
  '''
# ---
# name: TestStatsTablePreAggregated.test_country_breakdown_with_filters
  '''
  
  SELECT web_stats_daily.country_name AS `context.columns.breakdown_value`, tuple(uniqMergeIf(web_stats_daily.persons_uniq_state, and(greaterOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-02 23:59:59.999999', 6, 'UTC')))), uniqMergeIf(web_stats_daily.persons_uniq_state, and(greaterOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-02 23:59:59.999999', 6, 'UTC'))))) AS `context.columns.visitors`, tuple(sumMergeIf(web_stats_daily.pageviews_count_state, and(greaterOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-02 23:59:59.999999', 6, 'UTC')))), sumMergeIf(web_stats_daily.pageviews_count_state, and(greaterOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-02 23:59:59.999999', 6, 'UTC'))))) AS `context.columns.views` 
  FROM web_stats_daily FINAL 
  WHERE and(equals(web_stats_daily.team_id, 420), greaterOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-02 23:59:59.999999', 6, 'UTC')), ifNull(equals(web_stats_daily.browser, %(hogql_val_0)s), 0)) 
  GROUP BY `context.columns.breakdown_value` ORDER BY `context.columns.views` DESC 
  LIMIT 50000
  '''
# ---
# name: TestStatsTablePreAggregated.test_device_type_breakdown_query_generation
  '''
  
  SELECT web_stats_daily.device_type AS `context.columns.breakdown_value`, tuple(uniqMergeIf(web_stats_daily.persons_uniq_state, and(greaterOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-02 23:59:59.999999', 6, 'UTC')))), uniqMergeIf(web_stats_daily.persons_uniq_state, and(greaterOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-02 23:59:59.999999', 6, 'UTC'))))) AS `context.columns.visitors`, tuple(sumMergeIf(web_stats_daily.pageviews_count_state, and(greaterOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-02 23:59:59.999999', 6, 'UTC')))), sumMergeIf(web_stats_daily.pageviews_count_state, and(greaterOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-02 23:59:59.999999', 6, 'UTC'))))) AS `context.columns.views` 
  FROM web_stats_daily FINAL 
  WHERE and(equals(web_stats_daily.team_id, 420), greaterOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-02 23:59:59.999999', 6, 'UTC'))) 
  GROUP BY `context.columns.breakdown_value` ORDER BY `context.columns.views` DESC 
  LIMIT 50000
  '''
# ---
# name: TestStatsTablePreAggregated.test_fallback_to_regular_query_when_preaggregated_not_available
  '''
  
  SELECT breakdown_value AS `context.columns.breakdown_value`, tuple(uniq(filtered_person_id), NULL) AS `context.columns.visitors`, tuple(sum(filtered_pageview_count), NULL) AS `context.columns.views` 
  FROM (
  SELECT any(if(not(empty(events__override.distinct_id)), events__override.person_id, events.person_id)) AS filtered_person_id, count() AS filtered_pageview_count, replaceRegexpAll(nullIf(nullIf(JSONExtractRaw(events.properties, %(hogql_val_9)s), ''), 'null'), '^"|"$', '') AS breakdown_value, events__session.session_id AS session_id, any(events__session.`$is_bounce`) AS is_bounce, min(events__session.`$start_timestamp`) AS start_timestamp 
  FROM events LEFT JOIN (
  SELECT toString(reinterpretAsUUID(bitOr(bitShiftLeft(raw_sessions.session_id_v7, 64), bitShiftRight(raw_sessions.session_id_v7, 64)))) AS session_id, if(ifNull(equals(uniqMerge(raw_sessions.pageview_uniq), 0), 0), NULL, not(or(ifNull(greater(uniqMerge(raw_sessions.pageview_uniq), 1), 0), ifNull(greater(uniqMerge(raw_sessions.autocapture_uniq), 0), 0), greaterOrEquals(dateDiff(%(hogql_val_0)s, min(toTimeZone(raw_sessions.min_timestamp, %(hogql_val_1)s)), max(toTimeZone(raw_sessions.max_timestamp, %(hogql_val_2)s))), 10)))) AS `$is_bounce`, min(toTimeZone(raw_sessions.min_timestamp, %(hogql_val_3)s)) AS `$start_timestamp`, raw_sessions.session_id_v7 AS session_id_v7 
  FROM raw_sessions 
  WHERE and(equals(raw_sessions.team_id, 420), greaterOrEquals(fromUnixTimestamp(intDiv(toUInt64(bitShiftRight(raw_sessions.session_id_v7, 80)), 1000)), minus(assumeNotNull(toDateTime(%(hogql_val_4)s, %(hogql_val_5)s)), toIntervalDay(3))), lessOrEquals(fromUnixTimestamp(intDiv(toUInt64(bitShiftRight(raw_sessions.session_id_v7, 80)), 1000)), plus(assumeNotNull(toDateTime(%(hogql_val_6)s, %(hogql_val_7)s)), toIntervalDay(3)))) 
  GROUP BY raw_sessions.session_id_v7, raw_sessions.session_id_v7) AS events__session ON equals(toUInt128(accurateCastOrNull(events.`$session_id`, %(hogql_val_8)s)), events__session.session_id_v7) LEFT OUTER JOIN (
  SELECT argMax(person_distinct_id_overrides.person_id, person_distinct_id_overrides.version) AS person_id, person_distinct_id_overrides.distinct_id AS distinct_id 
  FROM person_distinct_id_overrides 
  WHERE equals(person_distinct_id_overrides.team_id, 420) 
  GROUP BY person_distinct_id_overrides.distinct_id 
  HAVING ifNull(equals(argMax(person_distinct_id_overrides.is_deleted, person_distinct_id_overrides.version), 0), 0) 
  SETTINGS optimize_aggregation_in_order=1) AS events__override ON equals(events.distinct_id, events__override.distinct_id) 
  WHERE and(equals(events.team_id, 420), and(or(and(greaterOrEquals(toTimeZone(events.timestamp, %(hogql_val_10)s), assumeNotNull(toDateTime(%(hogql_val_11)s, %(hogql_val_12)s))), lessOrEquals(toTimeZone(events.timestamp, %(hogql_val_13)s), assumeNotNull(toDateTime(%(hogql_val_14)s, %(hogql_val_15)s)))), 0), or(equals(events.event, %(hogql_val_16)s), equals(events.event, %(hogql_val_17)s)), 1, isNotNull(breakdown_value))) 
  GROUP BY session_id, breakdown_value) 
  GROUP BY `context.columns.breakdown_value` ORDER BY `context.columns.visitors` DESC, `context.columns.views` DESC, `context.columns.breakdown_value` ASC 
  LIMIT 50000
  '''
# ---
# name: TestStatsTablePreAggregated.test_initial_page_breakdown_with_bounce_rate_query_generation
  '''
  
  SELECT web_bounces_daily.entry_pathname AS `context.columns.breakdown_value`, tuple(uniqMergeIf(web_bounces_daily.persons_uniq_state, and(greaterOrEquals(web_bounces_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_bounces_daily.day_bucket, toDateTime64('2023-01-02 23:59:59.999999', 6, 'UTC')))), uniqMergeIf(web_bounces_daily.persons_uniq_state, and(greaterOrEquals(web_bounces_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_bounces_daily.day_bucket, toDateTime64('2023-01-02 23:59:59.999999', 6, 'UTC'))))) AS `context.columns.visitors`, tuple(sumMergeIf(web_bounces_daily.pageviews_count_state, and(greaterOrEquals(web_bounces_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_bounces_daily.day_bucket, toDateTime64('2023-01-02 23:59:59.999999', 6, 'UTC')))), sumMergeIf(web_bounces_daily.pageviews_count_state, and(greaterOrEquals(web_bounces_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_bounces_daily.day_bucket, toDateTime64('2023-01-02 23:59:59.999999', 6, 'UTC'))))) AS `context.columns.views`, tuple(divide(sumMergeIf(web_bounces_daily.bounces_count_state, and(greaterOrEquals(web_bounces_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_bounces_daily.day_bucket, toDateTime64('2023-01-02 23:59:59.999999', 6, 'UTC')))), nullIf(uniqMergeIf(web_bounces_daily.sessions_uniq_state, and(greaterOrEquals(web_bounces_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_bounces_daily.day_bucket, toDateTime64('2023-01-02 23:59:59.999999', 6, 'UTC')))), 0)), divide(sumMergeIf(web_bounces_daily.bounces_count_state, and(greaterOrEquals(web_bounces_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_bounces_daily.day_bucket, toDateTime64('2023-01-02 23:59:59.999999', 6, 'UTC')))), nullIf(uniqMergeIf(web_bounces_daily.sessions_uniq_state, and(greaterOrEquals(web_bounces_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_bounces_daily.day_bucket, toDateTime64('2023-01-02 23:59:59.999999', 6, 'UTC')))), 0))) AS `context.columns.bounce_rate` 
  FROM web_bounces_daily FINAL 
  WHERE and(equals(web_bounces_daily.team_id, 420), greaterOrEquals(web_bounces_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_bounces_daily.day_bucket, toDateTime64('2023-01-02 23:59:59.999999', 6, 'UTC'))) 
  GROUP BY `context.columns.breakdown_value` ORDER BY `context.columns.views` DESC 
  LIMIT 50000
  '''
# ---
# name: TestStatsTablePreAggregated.test_page_breakdown_with_bounce_rate_query_generation
  '''
  
  SELECT web_stats_daily.pathname AS `context.columns.breakdown_value`, tuple(uniqMergeIf(web_stats_daily.persons_uniq_state, and(greaterOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-02 23:59:59.999999', 6, 'UTC')))), uniqMergeIf(web_stats_daily.persons_uniq_state, and(greaterOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-02 23:59:59.999999', 6, 'UTC'))))) AS `context.columns.visitors`, tuple(sumMergeIf(web_stats_daily.pageviews_count_state, and(greaterOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-02 23:59:59.999999', 6, 'UTC')))), sumMergeIf(web_stats_daily.pageviews_count_state, and(greaterOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-02 23:59:59.999999', 6, 'UTC'))))) AS `context.columns.views`, any(bounces.`context.columns.bounce_rate`) AS `context.columns.bounce_rate` 
  FROM web_stats_daily FINAL LEFT JOIN (
  SELECT web_bounces_daily.entry_pathname AS `context.columns.breakdown_value`, tuple(uniqMergeIf(web_bounces_daily.persons_uniq_state, and(greaterOrEquals(web_bounces_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_bounces_daily.day_bucket, toDateTime64('2023-01-02 23:59:59.999999', 6, 'UTC')))), uniqMergeIf(web_bounces_daily.persons_uniq_state, and(greaterOrEquals(web_bounces_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_bounces_daily.day_bucket, toDateTime64('2023-01-02 23:59:59.999999', 6, 'UTC'))))) AS `context.columns.visitors`, tuple(sumMergeIf(web_bounces_daily.pageviews_count_state, and(greaterOrEquals(web_bounces_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_bounces_daily.day_bucket, toDateTime64('2023-01-02 23:59:59.999999', 6, 'UTC')))), sumMergeIf(web_bounces_daily.pageviews_count_state, and(greaterOrEquals(web_bounces_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_bounces_daily.day_bucket, toDateTime64('2023-01-02 23:59:59.999999', 6, 'UTC'))))) AS `context.columns.views`, tuple(divide(sumMergeIf(web_bounces_daily.bounces_count_state, and(greaterOrEquals(web_bounces_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_bounces_daily.day_bucket, toDateTime64('2023-01-02 23:59:59.999999', 6, 'UTC')))), nullIf(uniqMergeIf(web_bounces_daily.sessions_uniq_state, and(greaterOrEquals(web_bounces_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_bounces_daily.day_bucket, toDateTime64('2023-01-02 23:59:59.999999', 6, 'UTC')))), 0)), divide(sumMergeIf(web_bounces_daily.bounces_count_state, and(greaterOrEquals(web_bounces_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_bounces_daily.day_bucket, toDateTime64('2023-01-02 23:59:59.999999', 6, 'UTC')))), nullIf(uniqMergeIf(web_bounces_daily.sessions_uniq_state, and(greaterOrEquals(web_bounces_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_bounces_daily.day_bucket, toDateTime64('2023-01-02 23:59:59.999999', 6, 'UTC')))), 0))) AS `context.columns.bounce_rate` 
  FROM web_bounces_daily FINAL 
  WHERE equals(web_bounces_daily.team_id, 420) 
  GROUP BY `context.columns.breakdown_value`) AS bounces ON equals(web_stats_daily.pathname, bounces.`context.columns.breakdown_value`) 
  WHERE and(equals(web_stats_daily.team_id, 420), greaterOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-01 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-02 23:59:59.999999', 6, 'UTC'))) 
  GROUP BY `context.columns.breakdown_value` ORDER BY `context.columns.views` DESC 
  LIMIT 50000
  '''
# ---
# name: TestStatsTablePreAggregated.test_utm_source_breakdown_with_comparison_period
  '''
  
  SELECT web_stats_daily.utm_source AS `context.columns.breakdown_value`, tuple(uniqMergeIf(web_stats_daily.persons_uniq_state, and(greaterOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-08 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-15 23:59:59.999999', 6, 'UTC')))), uniqMergeIf(web_stats_daily.persons_uniq_state, and(greaterOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-08 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-15 23:59:59.999999', 6, 'UTC'))))) AS `context.columns.visitors`, tuple(sumMergeIf(web_stats_daily.pageviews_count_state, and(greaterOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-08 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-15 23:59:59.999999', 6, 'UTC')))), sumMergeIf(web_stats_daily.pageviews_count_state, and(greaterOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-08 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-15 23:59:59.999999', 6, 'UTC'))))) AS `context.columns.views` 
  FROM web_stats_daily FINAL 
  WHERE and(equals(web_stats_daily.team_id, 420), greaterOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-08 00:00:00.000000', 6, 'UTC')), lessOrEquals(web_stats_daily.day_bucket, toDateTime64('2023-01-15 23:59:59.999999', 6, 'UTC'))) 
  GROUP BY `context.columns.breakdown_value` ORDER BY `context.columns.views` DESC 
  LIMIT 50000
  '''
# ---
